text/0000-Component-Manager-Bounds.md

```diff
@@ -0,0 +1,177 @@
+
+## Motivation
+Historically, classic Ember components have had a built in escape hatch for doing DOM manipulation via `didInsertElement`. This allowed for components to encapsulate some 3rd party JavaScript library or do some fine grain DOM manipulation. We feel that is ability to drop down to a lower level abstraction makes components powerful.
+
+Today, the custom components are not given direct access to the DOM structure; internally known as `Bounds`. This limitation means that custom components cannot easily encapsulate functionality in the same way that classic Ember components can. Expanding the capabilities of custom components to expose the `Bounds` is in order to increase the usefulness of the abstraction.
+
+## Detailed design
+
+In [RFC#2013](https://github.com/emberjs/rfcs/blob/master/text/0213-custom-components.md#detailed-design) we outlined the component lifecycle and how a custom component is associated with a manager. This design builds off off of that existing work and concepts.
+
+### `elementHook` capability
+
+The primary way of introducing new functionality to the component manager is through capabilities. To allow custom component managers to announce that they support element related hooks, we will introduce an `elementHook` flag to the capabilities. In doing so, bounds will be dispatched to a `didRenderLayout` method on the manager and `willDestroyLayout` will be called during destruction.
+
+```js
+import { capabilities } from '@ember/component';
+import EmberObject from '@ember/object';
+
+export default EmberObject.extend({
+  capabilities: capabilities('3.5', {
+    elementHook: true
+  }),
+
+  // ...
+});
+```
+
+### `didRenderLayout` hook
+Once the rendering engine has constructed the bounds for the component the `didRenderLayout` hook will be called with the `component` and a `Bounds` instance.
+
+
+```js
+import { capabilities } from '@ember/component';
+import EmberObject from '@ember/object';
+
+export default EmberObject.extend({
+  capabilities: capabilities('3.5', {
+    elementHook: true
+  }),
+
+  didRenderLayout(component, bounds) {
+    component.bounds = bounds;
+    component.didInsertElement();
+  }
+  // ...
+});
+```
+
+#### Timing Semantics
+
+`didRenderLayout` has the following timing semantics:
+
+**Always**
+- called **before** `didCreateComponent`
+- called **after** the children's `didRenderLayout` hook is called
+- called **after** DOM insertion
+
+**May or May Not**
+- be called in the same tick as DOM insertion
+- be called in the same tick as `didCreateComponent`
+- have the the parent's children fully initialized in DOM
+
+In other words, components **should not** rely on the timing of DOM insertion outside of the component's own bounds.
+
+### `willDestroyLayout` hook
+
+To balance `didRenderLayout` we will expose a hook called `willDestroyLayout` that is invoked during destruction. It will receive the `component` instance.
```

Comment:
it doesn't get the component instance, it gets the "return value from createComponent" (may be a state bucket, may be a component instance)